app/cluster/state/etcd.go

```diff
@@ -0,0 +1,279 @@
+	Status string `json:"status"`
+}
+
+func EnvETCDConfig() *ETCDConfig {
+	endpoints := strings.Split(config.GetEnv("ETCD_HOSTS", "127.0.0.1:2379"), `,`)
+	releaseName := config.GetEnv("RELEASE_NAME", `multitenantv1`)
+	serverIndex := config.GetInstanceID()
+
+	envConfigOnce.Do(func() {
+		config.RegisterDurationConfigVariable(time.Duration(15), &etcdGetTimeout, true, time.Second, "etcd.getTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(3), &etcdWatchTimeout, true, time.Second, "etcd.watchTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(30), &keepaliveTime, true, time.Second, "etcd.keepaliveTime")
+		config.RegisterDurationConfigVariable(time.Duration(10), &keepaliveTimeout, true, time.Second, "etcd.keepaliveTimeout")
+		config.RegisterDurationConfigVariable(time.Duration(20), &dialTimeout, true, time.Second, "etcd.dialTimeout")
+	})
+
+	return &ETCDConfig{
+		Endpoints:            endpoints,
+		ReleaseName:          releaseName,
+		ServerIndex:          serverIndex,
+		etcdWatchTimeout:     etcdWatchTimeout,
+		dialTimeout:          dialTimeout,
+		dialKeepAliveTime:    keepaliveTime,
+		dialKeepAliveTimeout: keepaliveTimeout,
+		etcdGetTimeout:       etcdGetTimeout,
+	}
+}
+
+type ETCDManager struct {
+	Config  *ETCDConfig
+	Client  *clientv3.Client
+	once    sync.Once
+	initErr error
+}
+
+func (manager *ETCDManager) init() error {
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (manager *ETCDManager) prepareMode(raw []byte) servermode.ModeRequest {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ModeError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := servermode.Mode(req.Mode)
+	if !mode.Valid() {
+		return servermode.ModeError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewModeRequest(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return err
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ModeRequest {
+	ch := make(chan servermode.ModeRequest, 1)
+	ch <- servermode.ModeError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ModeRequest {
```

Comment:
Is this ever going to be used by more than one components/goroutines?
If the answer is no, return an error or panic the 2nd time, as a counter-measure for anybody trying to misuse it.
If the answer is yes, it wouldn't be efficient to create a new etcd watch every time a new consumer arrives. We can figure ways around this (typical Rx issue, we could use our existing `pubsub`), but I think the answer to the original question must be no, correct?

Reply:
, is `no`, but I find both proposals to be kind of an overkill. What we are trying to protect against, what is the benefit of adding this complexity?

Reply:
misuse, due to missing documentation